### PR TITLE
ref(mep): Move transaction thresholds to new tagging system, add histogram outliers [INGEST-541] [INGEST-947]

### PR DESCRIPTION
**Pull request by untitaker**
In https://github.com/getsentry/relay/pull/1225 we introduced a
general-purpose tagging system for transaction metrics. This may later
be extended:

* we will add outlier tagging for histograms soon
* session metrics could be tagged the same way (unlikely to happen for now, needs changes in Relay)

Here we move away from the purpose-built user satisfaction impl to the
more generic one

### CODE REVIEW

File: src/sentry/relay/config/metric_extraction.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    for threshold in project.projecttransactionthresholdoverride_set.all():
+        rules.extend(
+            _threshold_to_rules(
+                threshold,
+                [{"op": "eq", "name": "event.transaction", "value": threshold.transaction}],
+            )
+        )
+
+    try:
+        threshold = ProjectTransactionThreshold.objects.get(project=project)
```
Comment (jjbayer) [inline src/sentry/relay/config/metric_extraction.py:74]: Just to double check: When there are more than one `ProjectTransactionThreshold`s, the exception raised is caught by the catch-all on [line 200](https://github.com/getsentry/sentry/pull/33722/files#diff-5e30755537fcfd5687d4399272065e49affcc12a4cf7a0bc377cff5234b8a08aR200), right?

File: src/sentry/relay/config/metric_extraction.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    threshold=int(DEFAULT_THRESHOLD["threshold"]),
+)
+
+
+def get_metric_conditional_tagging_rules(
+    project: Project,
+) -> Sequence[MetricConditionalTaggingRule]:
+    rules: List[MetricConditionalTaggingRule] = []
+
+    for threshold in project.projecttransactionthresholdoverride_set.all():
```
Comment (jjbayer) [inline src/sentry/relay/config/metric_extraction.py:65]:
```suggestion
    
    # transaction-specific overrides must precede the project-wide threshold in the list of rules.
    for threshold in project.projecttransactionthresholdoverride_set.all():
```

File: src/sentry/relay/config/metric_extraction.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        "tagValue": "tolerated",
+    }
+    satisfied: MetricConditionalTaggingRule = {
+        "condition": {"op": "and", "inner": list(extra_conditions)},
+        "targetMetrics": list(_SATISFACTION_TARGET_METRICS),
+        "targetTag": _SATISFACTION_TARGET_TAG,
+        "tagValue": "satisfied",
+    }
+
+    return [frustrated, tolerated, satisfied]
```
Comment (jjbayer) [inline src/sentry/relay/config/metric_extraction.py:130]: Would also add a comment here that the order of rules is important.

File: src/sentry/relay/config/metric_extraction.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                    "op": "gt",
+                    "name": _TRANSACTION_METRICS_TO_RULE_FIELD[threshold.metric],
+                    # The frustration threshold is always four times the threshold
+                    # (see https://docs.sentry.io/product/performance/metrics/#apdex)
+                    "value": threshold.threshold * 4,
+                },
+                *extra_conditions,
+            ],
+        },
+        "targetMetrics": list(_SATISFACTION_TARGET_METRICS),
```
Comment (jjbayer) [inline src/sentry/relay/config/metric_extraction.py:103]: This seems to create flakey snapshots (see below). Maybe `sorted(...)` helps?

File: src/sentry/relay/config/metric_extraction.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/relay/config/metric_extraction.py:74]: yup I changed that behavior, there's a unique constraint on `project_id` on this table

**Review by jjbayer — APPROVED** [review]
Some minor remarks, otherwise LGTM!

File: src/sentry/relay/config/__init__.py [hunk 5]
```diff
@@ -190,6 +189,18 @@ def get_project_config(project, full_config=True, project_keys=None):
...
+        # This config key is technically not specific to _transaction_ metrics,
+        # is however currently both only applied to transaction metrics in
+        # Relay, and only used to tag transaction metrics in Sentry.
+        try:
+            cfg["config"]["metricConditionalTagging"] = get_metric_conditional_tagging_rules(
+                project
+            )
+        except Exception:
+            capture_exception()
+            raise
```
Comment (untitaker) [inline src/sentry/relay/config/__init__.py:202]:
```suggestion
```

File: src/sentry/relay/config/metric_extraction.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    ],
+    extra_conditions: Sequence[RuleConditionInner],
+) -> Sequence[MetricConditionalTaggingRule]:
+    frustrated: MetricConditionalTaggingRule = {
+        "condition": {
+            "op": "and",
+            "inner": [
+                {
+                    "op": "gt",
+                    "name": _TRANSACTION_METRICS_TO_RULE_FIELD[threshold.metric],
```
Comment (untitaker) [inline src/sentry/relay/config/metric_extraction.py:97]: fwiw this was ported from this piece of relay code: https://github.com/getsentry/relay/blob/043479fe570949003964c672ed3dda7253a98292/relay-server/src/metrics_extraction/transactions.rs#L118-L132=

**Comment by untitaker** [comment]
@jjbayer making you re-review this since I added histogram outliers

File: src/sentry/relay/config/metric_extraction.py [hunk 7]
```diff
@@ -116,15 +123,127 @@ def _threshold_to_rules(
...
+    rules = []
+    for platform, op, (p25, p75) in query_results:
+        rules.append(
+            {
+                "condition": {
+                    "op": "and",
+                    "inner": [
+                        {"op": "eq", "name": "event.contexts.trace.op", "value": op},
+                        {"op": "eq", "name": "event.platform", "value": platform},
+                        {"op": "gte", "name": "event.duration", "value": p25 + 3 * p75},
```
Comment (jjbayer) [inline src/sentry/relay/config/metric_extraction.py:226]: Could we add a comment here on why we use `p25 + 3 * p75`? I tried googling [Tukey's fences](https://en.wikipedia.org/wiki/Outlier#Tukey's_fences) but the only formula I found was `p75 + 3 * (p75 - p25)`.

File: src/sentry/relay/config/metric_extraction.py [hunk 8]
```diff
@@ -116,15 +123,127 @@ def _threshold_to_rules(
...
+                "tagValue": "outlier",
+            }
+        )
+
+    rules.append(
+        {
+            "condition": {
+                "op": "and",
+                "inner": [
+                    {"op": "gte", "name": "event.duration", "value": 0},
```
Comment (jjbayer) [inline src/sentry/relay/config/metric_extraction.py:240]: Does this mean that any `(platform, op)` combination that is not among the Top-50 gets tagged as 100% outliers?

File: src/sentry/relay/config/metric_extraction.py (hunk 7 above)
Comment (untitaker) [reply src/sentry/relay/config/metric_extraction.py:226]: oof you're right, the calculation is completely messed up. copied it from a document that is apparently wrong

**Review by jjbayer — APPROVED** [review]

**Pull request merged**